src/ir/module-splitting.cpp

```diff
@@ -971,11 +974,37 @@ void ModuleSplitter::removeUnusedSecondaryElements() {
   // code size in the primary module as well.
   for (auto& secondaryPtr : secondaries) {
     PassRunner runner(secondaryPtr.get());
+    // Do not validate here in the middle, as the IR still needs updating later.
+    runner.options.validate = false;
     runner.add("remove-unused-module-elements");
     runner.run();
   }
 }
 
+void ModuleSplitter::updateIR() {
+  // Imported functions may need type updates.
+  struct Fixer : public PostWalker<Fixer> {
+    void visitRefFunc(RefFunc* curr) {
+      auto& wasm = *getModule();
+      auto* func = wasm.getFunction(curr->func);
+      if (func->type != curr->type) {
+        // This became an import, and lost exactness.
+        assert(!func->type.isExact());
+        assert(curr->type.isExact());
+        if (wasm.features.hasCustomDescriptors()) {
+          // Add a cast, as the parent may depend on the exactness to validate.
+          replaceCurrent(Builder(wasm).makeRefCast(curr, curr->type));
+        }
```

Comment:
We don't have issues about needing this cast even when custom descriptors are disabled?

Reply:
For example in the test I commented on here: https://github.com/WebAssembly/binaryen/pull/7993#discussion_r2492371894. Shouldn't it be an IR validation error that we have an inexact reference being stored in an exact local? The binary writer would strip exactness and make the output valid, but I would have expected our validator to complain about the IR.

Reply:
Oh right, good point. Fixed and added a test to show we add a cast there too.

Reply:
Ok but now the problem is that exact casts are invalid unless custom descriptors is enabled :)